### ISSUE TITLE
Property diff

### ISSUE THREAD

**Issue by Matt-Esch** [Matt-Esch/virtual-dom#54]
Discussion on this has been neglected so I would like to solicit feedback on properties so that we have a mechanism for performing all operations

So the basic mechanism is

`diffProps({ test: "hello" }, { test: "goodbye"}) -> { test: "goodbye" }`

so basically, if 2 identical keys relate to different values, we patch the value

@Raynos tests suggest this should also work for undefined, but I am not 100% in agreement with this

``` javascript
diffProps({ test: "hello" }, { test: undefined })
// -> { test: undefined }`
```

Next we consider complex properties:

``` javascript
diffProps({ test: { a: 1, b: 2 } }, { test: { a: 1, b: 3 } })
// -> { test: { b: 3 } }
```

Note here that the diff produced excludes a, so a is left as it is.

Next we consider unset props

`diffProps({ a: 1, b: 2 }, { a : 1 }) -> {}`

So in this case, not setting b means "leave it as it is". This is where I would like to center the discussion since it's the most questionable. I believe this is valid. The shape of things to me is that all properties unspecified on a DOM node mean "leave it as the underlying value".

Hooks present us some trouble because if we want a hook to execute only once, there is currently no way to do this without setting the associated value to something or persisting the hook over multiple diffs.

If we accept my suggestion then `{ value: hook }` followed by `{}` means that the hook for value gets executed and then value is left alone. Furthermore, currently using `{ value: undefined }` means the same thing. This is particularly nice in ternary expressions `{ value : resetting ? resetValueHook("") : undefined }`

Generally speaking, all props tend to be enumerated, so the unset key is natural coding style. I think it's pretty valid to allow unset keys to remain untouched, I think we must consider if we use the value `undefined` to mean don't touch as well. Either we add a hook to mean don't touch, or if we use undefined to mean don't touch and add an undefined value hook.

So to make that clearer in code form, an execute hook once statement can either look like
- `{ value : resetting ? resetValueHook("") : undefined }`

Or we can define a noop hook
- `{ value : resetting ? resetValueHook("") : noop }`

Also, there is still the question of how we perform a delete operation. It seems reasonably valid to delete a key, but I would not imagine this is the norm. Hooks can do this, but perhaps we want something better?

Finally we need to consider if nested properties should be able to contain hooks. While it sounds like an extension of what we currently have, the "propName" value would be up for debate. Without nested hooks, and assuming we rely on hooks to delete keys, the only way to delete a key from a nested object would be to set the entire object with a hook.

To summarise the points that need clarifying:
- Do we agree that undefined keys should mean "don't touch"
- Do we use undefined to mean "don't touch" or expose a `noop` hook?
- Do we need a mechanism for deleting keys or will hooks suffice?
- Should hooks be limited to the root or should we allow them to be nested?
